book/src/proposals/durable-tx-nonces.md

```diff
@@ -0,0 +1,49 @@
+# Durable Transaction Nonces
+
+## Problem
+
+To prevent replay, Solana transactions contain a nonce field populated with a "recent" blockhash value. A transaction containing a blockhash that is too old (~2min as of this writing) is rejected by the network as invalid. Unfortunately certain use cases, such as custodial services, require a longer time allowance to produce a signature for the transaction. A mechanism is needed to enable these slow-signer network participants.
+
+## A Naive Solution
+
+An initial attempt at this was made by designating an special blockhash value that, if present in the `recent_blockhash` field, allowed the transaction to succeed so long as it additionally spent the full balance of the fee account.  However, this would introduce a security vulnerability by the fact that the transaction could be replayed by simply funding the fee account again.
+
+## A Contract-based Solution
+
+Here we propose a contract-based solution to the problem, whereby a client can "stash" a nonce value for future use in a transaction's `recent_blockhash` field. This approach is akin to the Compare and Swap atomic instruction, implemented by some CPU ISAs.
+
+### Contract Mechanics
+
+TODO: svgbob this into a flowchart, add prose
+
+```text
+Start
+Create Account
+  state = Uninitialized
+InitializeInstruction(hash)
+  if hash is recent
+    stored_hash = hash
+    state = Initialized
+    success
+  else
+    error
+SpendInstruction(spend_hash, next_hash)
```

Comment:
you don't need this instruction, you can use the semantics of "initialize", but maybe call it "Nonce", and what it always does is pick a new recent blockhash

Reply:
The nonce being consumed needs to be specified somewhere, no?

> but maybe call it "Nonce"

👍

Reply:
every time a "Nonce" instruction is issued, the nonce is simultaneously consumed and updated.  if the Nonce instruction cannot find a newer recent_blockhash than it currently carries, the instruction (and by extension the TX) fails

Reply:
Right.  What I'm getting at is that the stored hash needs to be specified somewhere in the TX in order to effect the signatures.  Leaving it in the `recent_blockhash` field would require exposing the TX message header to instructions and limit us to one instruction per TX, so passing it as a parameter to the instruction seems most useful.

Reply:
> Right. What I'm getting at is that the stored hash needs to be specified somewhere in the TX in order to effect the signatures.

every TX has a recent_blockhash that's covered by the signature

> Leaving it in the `recent_blockhash` field would require exposing the TX message header to instructions and limit us to one instruction per TX, so passing it as a parameter to the instruction seems most useful.

can you expand upon this?  I don't understand "would require exposing..." or "limit us to one instruction..."

Reply:
We need to verify that the hash the TX is specifying matches the stored one, no?  I didn't notice any exposure of the TX message header to the instruction processing path, but I didn't look that hard either.  I'll dig a little harder after lunch

RE limiting: If there were a case where multiple signers needed to use this feature, would they not have to coordinate stashing the same nonce in order to use the `recent_blockhash` field?

Reply:
@rob-solana did that clarify my thoughts any?